### ISSUE TITLE
Design Doc of how the schedules are being handled

### ISSUE THREAD

**Issue by shaikidris** [etf1/kafka-message-scheduler#19]
Hi Folks,

Appreciate if you could share the design doc on how the schedules are handled? i.e. how the service keeps track of schedules per event, is any other state store being used?

**Comment by fkarakas** [comment]
Hi @shaikidris,

The goal of the kafka message scheduler is to publish kafka messages to a specific topic with a specific id and payload in the future. Why do we need that ? well, to trigger some processing for example putting online or offline a video.

A schedule is just a kafka message with specific headers, and a payload. The schedule id should be uniq and if you want to change the payload of your message, you just have to publish a new version of your message in the scheduler topic. The topic can have many versions fo your schedules, the scheduler will always take the latest. There is a client lib in go for preparing a kafka message to be a schedule: https://github.com/etf1/kafka-message-scheduler/tree/main/clientlib

Basically, the scheduler reads the topic from the beginning at startup and every day at 00:00 am and it stores in memory the schedules planned for the current day. It will also watch new incoming messages and process only messages for the current by adding, updating or deleting schedules in the memory store.

So there is no other storage than the scheduler kafka topic, no external db etc... All schedules are stored in a kafka topic.

I hope it is clear enough, don't hesitate to ask us for any further information.

**Comment by shaikidris** [comment]
Thanks @fkarakas.

That helps, pardon my ignorance of not getting in to the go code.
My usecase is to handle millions of scheduled events per day with varied time (i.e. each event can have a unique schedule time). 
With the tombstone approach won't the schedule topic keeps on growing?

**Comment by fkarakas** [comment]
You can fine-tune the scheduler topic depending on your needs. If the topic is compacted, the tombstone message and all messages with the same id will be deleted from the topic at some point when the kafka log compacter will be triggered based on some thresholds like segment or retention etc ...

if you want to be able to trigger millions of events per day, perhaps the best thing is to partition the topic so many instances of the scheduler will work together by splitting the work, each scheduler handling a part of the events. Otherwise with one scheduler, the memory pressure can be large depending of course of the size of your kafka messages.

We never pushed the scheduler as far with millions of events per day, but if you encounter issues don't hesitate to share.

**Comment by fkarakas** [comment]
closing for now, if necessary you can reopen
